server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -81,6 +102,102 @@ func (b *BuildEventHandler) OpenChannel(ctx context.Context, iid string) interfa
 		hasReceivedStartedEvent: false,
 		eventsBeforeStarted:     make([]*inpb.InvocationEvent, 0),
 		logWriter:               logWriter,
+		onClose:                 onClose,
+	}
+}
+
+func (b *BuildEventHandler) GetStatsRecorder() *statsRecorder {
+	return b.statsRecorder
+}
+
+// recordStatsRequest is a request to record the stats for an invocation, which
+// will be fulfilled by one of the workers spawned by StatsRecorder.
+type recordStatsRequest struct {
+	invocationID string
+	requestedAt  time.Time
+}
+
+// statsRecorder listens for finalized invocations and copies cache stats from
+// the metrics collector to the DB.
+type statsRecorder struct {
+	env          environment.Env
+	openChannels *sync.WaitGroup
+	workerDone   chan struct{}
+
+	mu      sync.Mutex // protects(stopped, reqs)
+	reqs    chan *recordStatsRequest
+	stopped bool
+}
+
+func newStatsRecorder(env environment.Env, openChannels *sync.WaitGroup) *statsRecorder {
+	return &statsRecorder{
+		env:          env,
+		openChannels: openChannels,
+		reqs:         make(chan *recordStatsRequest, 4096),
+		workerDone:   make(chan struct{}),
+	}
+}
+
+func (r *statsRecorder) MarkFinalized(invocationID string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stopped {
+		alert.UnexpectedEvent(
+			"stats_recorder_finalize_after_shutdown",
+			"Invocation %q was marked finalized after the stats recorder was shut down.",
+			invocationID)
+	}
+
+	req := &recordStatsRequest{
+		invocationID: invocationID,
+		requestedAt:  time.Now(),
+	}
+	r.reqs <- req
```

Comment:
This will block if r.reqs is full. You should explicitly handle that case, maybe by logging an warning and dropping the invocationID?

Reply:
Since this should only ever happen when the DB is under such a high load that we can't flush stats fast enough, the blocking behavior here seems roughly equivalent to the old behavior -- `FinalizeInvocation` would effectively block until the DB load lightened up and the invocation row was able to be written. And since the behavior is more or less the same either way, I'm leaning towards blocking rather than dropping the stats, since dropping the stats seems worse for the user. wdyt?

Reply:
Resolved offline -- decided that since this blocks all builds from finishing and there might be other unexpected reasons why the buffer fills up, we'd rather just log an error.